### PR TITLE
executor: support mergeSort different selectResult in TableScan and IndexScan

### LINKED ISSUE #26166: Push down sort property to range partition table
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
mysql> create table t1(id int primary key, v int) partition by range(id) (partition p1 values less than(10000), partition p2 values less than(20000));
Query OK, 0 rows affected (0.01 sec)

mysql> create table t2(id int primary key, v int);
Query OK, 0 rows affected (0.01 sec)

mysql> explain select * from t1 where id > 20001 order by id limit 1;
+-----------------------------+---------+-----------+----------------+----------------------------------------------------+
| id                          | estRows | task      | access object  | operator info                                      |
+-----------------------------+---------+-----------+----------------+----------------------------------------------------+
| TopN_8                      | 1.00    | root      |                | test.t1.id, offset:0, count:1                      |
| └─TableReader_17            | 1.00    | root      | partition:dual | data:TopN_16                                       |
|   └─TopN_16                 | 1.00    | cop[tikv] |                | test.t1.id, offset:0, count:1                      |
|     └─TableRangeScan_15     | 3333.33 | cop[tikv] | table:t1       | range:(20001,+inf], keep order:false, stats:pseudo |
+-----------------------------+---------+-----------+----------------+----------------------------------------------------+
4 rows in set (0.00 sec)

mysql> explain select * from t2 where id > 20001 order by id limit 1;
+-----------------------------+---------+-----------+---------------+---------------------------------------------------+
| id                          | estRows | task      | access object | operator info                                     |
+-----------------------------+---------+-----------+---------------+---------------------------------------------------+
| Limit_12                    | 1.00    | root      |               | offset:0, count:1                                 |
| └─TableReader_22            | 1.00    | root      |               | data:Limit_21                                     |
|   └─Limit_21                | 1.00    | cop[tikv] |               | offset:0, count:1                                 |
|     └─TableRangeScan_20     | 1.00    | cop[tikv] | table:t2      | range:(20001,+inf], keep order:true, stats:pseudo |
+-----------------------------+---------+-----------+---------------+---------------------------------------------------+
4 rows in set (0.01 sec)

mysql> set tidb_partition_prune_mode='dynamic';
Query OK, 0 rows affected (0.00 sec)

mysql> explain select * from t1 where id > 20001 order by id limit 1;
+-----------------------------+---------+-----------+----------------+----------------------------------------------------+
| id                          | estRows | task      | access object  | operator info                                      |
+-----------------------------+---------+-----------+----------------+----------------------------------------------------+
| TopN_8                      | 1.00    | root      |                | test.t1.id, offset:0, count:1                      |
| └─TableReader_17            | 1.00    | root      | partition:dual | data:TopN_16                                       |
|   └─TopN_16                 | 1.00    | cop[tikv] |                | test.t1.id, offset:0, count:1                      |
|     └─TableRangeScan_15     | 3333.33 | cop[tikv] | table:t1       | range:(20001,+inf], keep order:false, stats:pseudo |
+-----------------------------+---------+-----------+----------------+----------------------------------------------------+
4 rows in set (0.00 sec)

mysql> explain select * from t2 where id > 20001 order by id limit 1;
+-----------------------------+---------+-----------+---------------+---------------------------------------------------+
| id                          | estRows | task      | access object | operator info                                     |
+-----------------------------+---------+-----------+---------------+---------------------------------------------------+
| Limit_12                    | 1.00    | root      |               | offset:0, count:1                                 |
| └─TableReader_22            | 1.00    | root      |               | data:Limit_21                                     |
|   └─Limit_21                | 1.00    | cop[tikv] |               | offset:0, count:1                                 |
|     └─TableRangeScan_20     | 1.00    | cop[tikv] | table:t2      | range:(20001,+inf], keep order:true, stats:pseudo |
+-----------------------------+---------+-----------+---------------+---------------------------------------------------+
4 rows in set (0.00 sec)
```

### 2. What did you expect to see? (Required)

t1, t2 be same

### 3. What did you see instead (Required)

the partition table can not be eliminated as limitN (even with dynamic prune)

### 4. What is your TiDB version? (Required)

newest master

### LINKED ISSUE #41801: the subquery result is wrong
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
source  [test.txt](https://github.com/pingcap/tidb/files/10856845/test.txt)
# maybe wait for hours
```

### 2. What did you expect to see? (Required)
```
MySQL [test]> select col1, col2 from UK_HP16709 where col1 > (select max(col1) from UK_HP16709 t1 where t1.col1 > 1113867879);
Empty set (0.06 sec)
```
### 3. What did you see instead (Required)
```
mysql> select col1, col2 from UK_HP16709 where col1 > (select max(col1) from UK_HP16709 t1 where t1.col1 > 1113867879);
+------------+--------------------------------------------------------------+
| col1       | col2                                                         |
+------------+--------------------------------------------------------------+
| 3747335913 | 澸穒艍朇鶩襽状闕璚筴甡堃鑽眐榯藖禬呈擰讳                     |
| 4270433610 | 抧銲勇轰贂勦螜婍唐藮冪瞈卞袈碗逗鼌鶯揬澳                     |
| 4289373785 | 绅讋檾郊鋱杹咋籲堲县忱鐯鉵凃秚袖雱鹢项臵                     |
| 4254321426 | 焯紈枎硬颿闿遛昢楑啲瀸礐橷灬齾鮚笂侠敓忳                     |
| 4257249400 | 鏜煜鉮絇詳挌而唆貋糥導撚菛砾Ĉ覩獦笊挤戤                      |
| 4016492889 | 儭章程顎菑竹磟稃琷漩弄扏煒鶈喌狳檠釈欉ǲ                      |
| 3722399782 | 男慦羙圲痼蝺č冣鼙颪銝棨锚媴桵紇釾傪ᛋ瀥                       |
| 3625863817 | 賰皲孎侐钬珩靹搱灂驿诶榧抚偘捚盷恨容磍黵                     |
| 4250099741 | 琣置鐐砛搹ň蠮瑇檙鍢潂賱鏩偵枖賸唱刨宇唬                      |
| 4141088841 | 鼀傝颬媾蚝杫率褞瞪愔朲鴏輅椰遪饯琤奂椼纈                     |
| 4047043168 | 妫壼紤乼罌犏鲰郐椋t畝湹昦苻悴耬翄ĸ累婑                       |
| 3955035493 | Ɉ堼殺儭鮺虐嚜Ͱ包獚垙悷鹾蟍銢漌笕ǈ嚲扰                        |
| 4294967295 | 炚欠嘔戭濧鸲铼皿釔梋呯舮嬘芵畫黵剝璅鑰瞖                     |
| 3847101422 | 毖铏憛唝刭傣骝絁\顏荍眨鹅谗喼俑窫擅淇猂                      |
+------------+--------------------------------------------------------------+
14 rows in set (0.05 sec)
```
### 4. What is your TiDB version? (Required)

```
Release Version: v6.7.0-alpha
Edition: Community
Git Commit Hash: 61ed093fe1f47911fcb2b1a8ab1f33be0cc5aaca
Git Branch: heads/refs/tags/v6.7.0-alpha
UTC Build Time: 2023-02-28 07:23:38
GoVersion: go1.20.1
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv
```

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #41801, ref #26166

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: distsql/select_result.go [hunk 1]
```diff
@@ -74,6 +77,120 @@ type SelectResult interface {
...
+				if ssr.cachedChunks[i].NumRows() == 0 {
+					continue
+				}
+			}
+			idx := ssr.cachedChunkIdx[i]
+			row := ssr.cachedChunks[i].GetRow(idx)
+			if minSelectResultIdx == -1 || ssr.lessRow(row, ssr.cachedChunks[minSelectResultIdx].GetRow(ssr.cachedChunkIdx[minSelectResultIdx])) {
+				minSelectResultIdx = i
+			}
+		}
```
Comment (winoros) [inline distsql/select_result.go:172]: We can use a heap to avoid the O(nm) algo.

File: executor/partition_table.go [hunk 2]
```diff
@@ -23,6 +23,9 @@ import (
 )
 
 func updateExecutorTableID(ctx context.Context, exec *tipb.Executor, recursive bool, partitionIDs []int64) error {
+	if exec == nil {
+		return nil
+	}
```
Comment (winoros) [inline executor/partition_table.go:28]: we enter here and caused panic?

Comment (Defined2014) [reply executor/partition_table.go:28]: Yes, `exec != nil` when `e.storeType == kv.TiFlash`.

File: executor/distsql.go [hunk 3]
```diff
@@ -324,28 +345,37 @@ func (e *IndexReaderExecutor) open(ctx context.Context, kvRanges []kv.KeyRange)
...
-	kvReq, err := builder.Build()
-	if err != nil {
-		e.feedback.Invalidate()
-		return err
-	}
-	e.result, err = e.SelectResult(ctx, e.ctx, kvReq, retTypes(e), e.feedback, getPhysicalPlanIDs(e.plans), e.id)
-	if err != nil {
-		e.feedback.Invalidate()
-		return err
+	if e.byItems == nil {
```
Comment (winoros) [inline executor/distsql.go:348]: `e.byItems == nil || e is not partition table`?

Comment (Defined2014) [reply executor/distsql.go]: Done

**Review by winoros — APPROVED** [review]
It's a very clean implementation for me!
Except that golang doesn't provide a generic heap implementation in std lib yet.

**Review by XuHuaiyu — APPROVED** [review]

File: executor/distsql.go [hunk 4]
```diff
@@ -294,6 +297,25 @@ func (e *IndexReaderExecutor) Open(ctx context.Context) error {
...
+		SetDesc(e.desc).
+		SetKeepOrder(e.keepOrder).
+		SetTxnScope(e.txnScope).
+		SetReadReplicaScope(e.readReplicaScope).
+		SetIsStaleness(e.isStaleness).
+		SetFromSessionVars(e.ctx.GetSessionVars()).
+		SetFromInfoSchema(e.ctx.GetInfoSchema()).
+		SetMemTracker(e.memTracker).
+		SetClosestReplicaReadAdjuster(newClosestReadAdjuster(e.ctx, &builder.Request, e.netDataSize)).
+    SetConnID(e.ctx.GetSessionVars().ConnectionID)
```
Comment (Defined2014) [inline executor/distsql.go:314]:
```suggestion
        SetConnID(e.ctx.GetSessionVars().ConnectionID)
```

**Pull request merged**